plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -927,30 +1011,72 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 		}
 
 		self->read_first_bank =
-		    (self->isp_model == ISP_MODEL_HUB_GL3523) && self->fw_bank_vers[0] != 0;
+		    (self->chip.model == ISP_MODEL_HUB_GL3523) && self->fw_bank_vers[0] != 0;
 		self->write_recovery_bank = address == self->fw_bank_addr[1];
 	}
 
 	/* has public key */
 	if (fu_device_has_private_flag(device, FU_GENESYS_USBHUB_FLAG_HAS_PUBLIC_KEY)) {
 		g_autofree gchar *guid = NULL;
-		if (!fu_memcpy_safe(self->public_key,
-				    sizeof(self->public_key),
-				    0, /* dst */
-				    g_bytes_get_data(blob, NULL),
-				    g_bytes_get_size(blob),
-				    self->fw_data_total_count, /* src */
-				    sizeof(self->public_key),
-				    error))
+		if (!fu_genesys_usbhub_device_read_flash(self,
+							 self->fw_key_addr[0],
+							 self->public_key,
+							 sizeof(self->public_key),
+							 NULL,
+							 error))
 			return FALSE;
 		guid = fwupd_guid_hash_data(self->public_key,
 					    sizeof(self->public_key),
 					    FWUPD_GUID_FLAG_NONE);
 		fu_device_add_instance_strup(device, "PUBKEY", guid);
+		if (!fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL))
+			return FALSE;
+	} else {
```

Comment:
I guess it is fair to have the following instance id for device that support public-key; i.e. to put the following change not under the else statement.

``` diff
                fu_device_add_instance_strup(device, "PUBKEY", guid);
        }
+
+        /* add specific product info */
+        if (self->running_bank != BANK_MASK_CODE) {
+ ...
+        }

+        fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "IC", NULL);
+        fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "IC", "BONDING", NULL);
+        fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "VENDOR", IC", "BONDING", "PORTNUM", "VENDORSUP", NULL);
        fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL);
```

Reply:
`PUBKEY` cannot be set without `FU_GENESYS_USBHUB_FLAG_HAS_PUBLIC_KEY` set.
How about add `fu_genesys_usbhub_build_extend_instance_id()`?

Reply:
You can just ignore the return value, as it was.

Reply:
The `README.md` should be updated accordingly with the new introduced instance id.

Reply:
> You can just ignore the return value, as it was.

I saw that :). It's interesting to check, in somewhere this "must" to do. I will modify on next.

---
BTW, @hughsie I found an error in `fu_usb_device_query_hub()` https://github.com/fwupd/fwupd/blob/d2adec18ca6838bfc0312fbe6b02983af680ac00/libfwupdplugin/fu-usb-device.c#L200-L207

The [USB 2.0 Specification](https://www.usb.org/document-library/usb-20-specification) `11.23.2.1 Hub Descriptor` described the `bDescLength` should fulfill `bNbrPorts`.

When I test our 8 ports GL3590, it was rejected in `fu_usb_device_setup()` https://github.com/fwupd/fwupd/blob/d2adec18ca6838bfc0312fbe6b02983af680ac00/libfwupdplugin/fu-usb-device.c#L324-L328

Reply:
@adamgene would this be more accurate?

	if (sz >= 0x0C) {
		g_autofree gchar *hub = g_strdup_printf("%02X%02X", data[0x0B], data[0x0A]);
		fu_device_add_instance_str(FU_DEVICE(self), "HUB", hub);
	} else if (sz >= 0x08) {
		fu_device_add_instance_u8(FU_DEVICE(self), "HUB", data[0x07]);
	}
	fu_device_build_instance_id(FU_DEVICE(self), NULL, "VID", "PID", "HUB", NULL)

Reply:
In USB 3 hub, the sz is fixed 0x0C(because max supporting ports are 15).
In USB 2 hub, the sz is in 0x09~(because max supporting ports are 255).

My 8 ports GL3590 replied as below. Hope this is a good info for you.

- Replied from USB 2
```
        ------------------- USB Hub Descriptor -----------------
bDescriptorLength        : 0x0A (10 bytes)
bDescriptorType          : 0x29 (Hub Descriptor)
bNumberOfPorts           : 0x08 (8 Ports)
wHubCharacteristics      : 0x00E9
 Power switching         : Individual
 Compound device         : No
 Over-current protection : Individual
 TT Think Time           : At most 32 FS bit times
 Port Indicators         : Supported
bPowerOnToPowerGood      : 0x3C (120 ms)
bHubControlCurrent       : 0x64 (100 mA)
DeviceRemovable          : 0x00, 0x00
 Bit 0                   : 0 (reserved, any value)
 Bit 1                   : 0 (Device at Port 1 is removable)
 Bit 2                   : 0 (Device at Port 2 is removable)
 Bit 3                   : 0 (Device at Port 3 is removable)
 Bit 4                   : 0 (Device at Port 4 is removable)
 Bit 5                   : 0 (Device at Port 5 is removable)
 Bit 6                   : 0 (Device at Port 6 is removable)
 Bit 7                   : 0 (Device at Port 7 is removable)
 Bit 8                   : 0 (Device at Port 8 is removable)
 Bit 9                   : 0 (unused, must be 0)
 Bit10                   : 0 (unused, must be 0)
 Bit11                   : 0 (unused, must be 0)
 Bit12                   : 0 (unused, must be 0)
 Bit13                   : 0 (unused, must be 0)
 Bit14                   : 0 (unused, must be 0)
 Bit15                   : 0 (unused, must be 0)
PowerControlMask         : 0xFF, 0x00
Data (HexDump)           : 0A 29 08 E9 00 3C 64 00 00 FF
```

- Replied from USB 3
```
        ------------------- USB Hub Descriptor -----------------
bDescriptorLength        : 0x09 (9 bytes)
bDescriptorType          : 0x29 (Hub Descriptor)
bNumberOfPorts           : 0x04 (4 Ports)
wHubCharacteristics      : 0x0009
 Power switching         : Individual
 Compound device         : No
 Over-current protection : Individual
 TT Think Time           : At most 8 FS bit times
 Port Indicators         : Not supported
bPowerOnToPowerGood      : 0x3C (120 ms)
bHubControlCurrent       : 0x90 (144 mA)
DeviceRemovable          : 0x00
 Bit 0                   : 0 (reserved, any value)
 Bit 1                   : 0 (Device at Port 1 is removable)
 Bit 2                   : 0 (Device at Port 2 is removable)
 Bit 3                   : 0 (Device at Port 3 is removable)
 Bit 4                   : 0 (Device at Port 4 is removable)
 Bit 5                   : 0 (unused, must be 0)
 Bit 6                   : 0 (unused, must be 0)
 Bit 7                   : 0 (unused, must be 0)
Data (HexDump)           : 09 29 04 09 00 3C 90 00 FF

        ---------------- USB 3.0 Hub Descriptor ---------------
bLength                  : 0x0C (12 bytes)
bDescriptorType          : 0x2A
bNumberOfPorts           : 0x04 (4 Ports)
wHubCharacteristics      : 0x0009
 Power switching         : Individual
 Compound device         : No
 Over-current protection : Individual
 TT Think Time           : At most 8 FS bit times
 Port Indicators         : Not supported
bPowerOnToPowerGood      : 0x3C (120 ms)
bHubControlCurrent       : 0x90 (576 mA)
bHubHdrDecLat            : 0x04 (0.4 µs)
wHubDelay                : 0x190 (400 ns)
DeviceRemovable          : 0x00
 Bit 0                   : 0 (reserved, any value)
 Bit 1                   : 0 (Device at Port 1 is removable)
 Bit 2                   : 0 (Device at Port 2 is removable)
 Bit 3                   : 0 (Device at Port 3 is removable)
 Bit 4                   : 0 (Device at Port 4 is removable)
 Bit 5                   : 0 (unused, must be 0)
 Bit 6                   : 0 (unused, must be 0)
 Bit 7                   : 0 (unused, must be 0)
Data (HexDump)           : 0C 2A 04 09 00 3C 90 04 90 01 00 00
```

Reply:
TIL: https://github.com/fwupd/fwupd/pull/4446 -- does that work?